src/test_utils.rs

```diff
@@ -29,49 +31,110 @@ use crate::dc_receive_imf::dc_receive_imf;
+    }
+
+    /// Configures as bob@example.net with fixed secret key.
+    ///
+    /// This is a shortcut for `.with_key_pair(bob_keypair()).
+    pub fn configure_bob(self) -> Self {
+        self.with_key_pair(bob_keypair())
+    }
+
+    /// Configures the new [`TestContext`] with the provided [`KeyPair`].
+    ///
+    /// This will extract the email address from the key and configure the context with the
+    /// given identity.
+    pub fn with_key_pair(mut self, key_pair: KeyPair) -> Self {
+        self.key_pair = Some(key_pair);
+        self
+    }
+
+    /// Attaches a [`LogSink`] to this [`TestContext`].
+    ///
+    /// This is useful when using multiple [`TestContext`] instances in one test: it allows
+    /// using a single [`LogSink`] for both contexts.  This shows the log messages in
+    /// sequence as they occurred rather than all messages from each context in a single
+    /// block.
+    pub fn with_log_sink(mut self, sink: Sender<Event>) -> Self {
+        self.log_sink = Some(sink);
+        self
+    }
+
+    /// Builds the [`TestContext`].
+    pub async fn build(self) -> TestContext {
+        let name = self.key_pair.as_ref().map(|key| key.addr.local.clone());
+
+        let test_context = TestContext::new_internal(name, self.log_sink).await;
+
+        if let Some(key_pair) = self.key_pair {
+            test_context
+                .configure_addr(&key_pair.addr.to_string())
+                .await;
+            key::store_self_keypair(&test_context, &key_pair, KeyPairUse::Default)
+                .await
+                .expect("Failed to save key");
+        }
+        test_context
+    }
+}
+
 /// A Context and temporary directory.
 ///
 /// The temporary directory can be used to store the SQLite database,
 /// see e.g. [test_context] which does this.
-pub(crate) struct TestContext {
+#[derive(Debug)]
+pub struct TestContext {
     pub ctx: Context,
     pub dir: TempDir,
-    pub evtracker: EvTracker,
+    pub evtracker: EventTracker,
     /// Counter for fake IMAP UIDs in [recv_msg], for private use in that function only.
     recv_idx: RwLock<u32>,
-    /// Functions to call for events received.
-    event_sinks: Arc<RwLock<Vec<Box<EventSink>>>>,
+    /// Channels which should receive events from this context.
+    event_senders: Arc<RwLock<Vec<Sender<Event>>>>,
     /// Receives panics from sinks ("sink" means "event handler" here)
-    poison_receiver: channel::Receiver<String>,
+    poison_receiver: Receiver<String>,
+    /// Reference to implicit [`LogSink`] so it is dropped together with the context.
+    ///
+    /// Only used if no explicit [`LogSink`] was given during construction.  This is a
+    /// convenience in case only a single [`TestContext`] is used to avoid dealing with
+    /// [`LogSink`].  Never read, thus "dead code", since the only purpose is to control
+    /// when Drop is invoked..
+    #[allow(dead_code)]
+    log_sink: Option<LogSink>,
```

Comment:
Not sure about this approach, it has a really high cognitive complexity ("if there is an explicit LogSink, then there is not a LogSink here, but somewhere else, namely directly in the test, but we always have a Sender for the logs, which is not visible here because it goes into the `event_senders`. A LogSink is the thing that receives the events and prints them at the end, while the Sender is the thing with which you can `send()` the events when you get them from the core.").

My alternative isn't great either, but still better I think. Or maybe we find yet another alternative. Or if not some comment that explains the bigger picture could help, I could also help with writing this in this case.

My alternative:

Remove `log_sink` here. If there is an explicit log sink given, don't change anything.
If there is no explicit log sink given, just print the events as we did before.

We should probably make the new log sink sender an explicit `Option<Sender<Event>>` then instead of putting it into `event_senders` in order to still know whether we have an explicit log sink or not.

(my approach will also has the disadvantage that when running multiple tests the output will be as splattered as before this PR, and the advantage that when running a single test the output will be as nice as before  - the splattering never was a problem for me personally as I rarely run multiple tests at once while debugging, so maybe I'm biased).

--

The sentence 

> Only used if no explicit [`LogSink`] was given during construction

is technically wrong and confused me when reading. Technically correct (but not a helpful comment) would be 

> Only used if 'log_sender: Option<Sender<Event>>' was `None` during construction'

note that the caller doesn't provide a `LogSink` but a `Sender<Event>`.

Reply:
> Not sure about this approach, it has a really high cognitive complexity ("if there is an explicit LogSink, then there is not a LogSink here, but somewhere else, namely directly in the test, but we always have a Sender for the logs, which is not visible here because it goes into the `event_senders`. A LogSink is the thing that receives the events and prints them at the end, while the Sender is the thing with which you can `send()` the events when you get them from the core.").
> 
> My alternative isn't great either, but still better I think. Or maybe we find yet another alternative. Or if not some comment that explains the bigger picture could help, I could also help with writing this in this case.
> 
> My alternative:
> 
> Remove `log_sink` here. If there is an explicit log sink given, don't change anything. If there is no explicit log sink given, just print the events as we did before.
> 
> We should probably make the new log sink sender an explicit `Option<Sender<Event>>` then instead of putting it into `event_senders` in order to still know whether we have an explicit log sink or not.
> 

> (my approach will also has the disadvantage that when running multiple tests the output will be as splattered as before this PR, and the advantage that when running a single test the output will be as nice as before - the splattering never was a problem for me personally as I rarely run multiple tests at once while debugging, so maybe I'm biased).

So yes, this is kind of the entire crux of this PR as you point out here.  Without an implicit LogSink the output will be splattered and highly confusing (to me at least, made worse by the fact that if you have `test_spam` and `test_spam_and_ham` it seems impossible to only run `test_spam` so when both are failing your test output is useless).

The point here is that by creating the LogSink implicitly the vast majority of the tests do not have to care or be changed at all to start behaving correctly.  And also **all** tests will have their logs contained in their own test avoiding leaking to other tests.  We only need to change tests which use multiple TestContext's and where we want the output interleaved.

> The sentence
> 
> > Only used if no explicit [`LogSink`] was given during construction
> 
> is technically wrong and confused me when reading. Technically correct (but not a helpful comment) would be
> 
> > Only used if 'log_sender: Option<Sender>' was `None` during construction'
> 
> note that the caller doesn't provide a `LogSink` but a `Sender<Event>`.

So I tried to improve the comments, hopefully this makes it less confusing and also reduces the cognitive complexity.  I really would like to avoid degrading the behaviour here because it is explicitly the main goal of this PR.

Reply:
> if you have `test_spam` and `test_spam_and_ham` it seems impossible to only run `test_spam`

You can use e.g. `cargo test -- dc_tools::tests::test_parse_receive_headers --exact` to run `test_parse_receive_headers` but not `test_parse_receive_headers_integration`. However, of course that's quite a lot to type just to run one test and I'm personally usually using the "Run" button which rust-analyzer adds to VSCode instead in this case because I don't want to type all of this.

The comments are a lot better now :+1:

Reply:
TIL about `--exact`, thanks (i suspected this must be possible somehow)

Reply:
ah, i'm not yet used to rust-analyser's adding of those lens things.  they also work in emacs and indeed use `--exact`